### DISCUSSION TITLE
How to deploy to other Linux distributions

### DISCUSSION THREAD

**Discussion by 2gud4u** [raysan5/raylib#2789] [Q&A] [answered]
Hello community.-

I compile/link my game with the command gcc -o exename main.c -lraylib -lstdc++ on Linux, but i have sent my "exe" to another Linux user and he says he cant run it. Could someone tell me what im doing wrong in order for my "exe" to run in other Linux dristributions ? Thanks in advance !!!

**Comment by chriscamacho** [comment]
i generally just statically compile raylib, its not a gigantic library...

On Sat, 5 Nov 2022 at 20:10, Jorge H Carretero ***@***.***>
wrote:

> Hello community.-
>
> I compile/link my game with the command gcc -o exename main.c -lraylib
> -lstdc++ on Linux, but i have sent my "exe" to another Linux user and he
> says he cant run it. Could someone tell me what im doing wrong in order for
> my "exe" to run in other Linux dristributions ? Thanks in advance !!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/raysan5/raylib/discussions/2789>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFO4SOL5CNRC5QZEAEFHE3WG25MRANCNFSM6AAAAAARYCL6TE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 

blog bedroomcoders.co.uk <http://bedroomcoders.co.uk>
Free source code and tutorials!


Disclaimer:
By sending an email to ANY of my addresses you are agreeing that:

   1. I am by definition, "the intended recipient"

   2. All information in the email is mine to do with as I see fit and make
such financial profit, political mileage, or good joke as it lends itself
to. In particular, I may quote it where I please.

   3. I may take the contents as representing the views of your company.

   4. This overrides any disclaimer or statement of confidentiality that
may be included on your message.

**Reply by 2gud4u** [reply]
Thank you !! could you provide a link or example on how to do that ? I got raylib from Linux os packages

**Comment by gen2brain** [comment]
@2gud4u The raylib by default is compiled static, i.e. you get `libraylib.a` that you use to link your app/game. Not sure how Cmake builds work, probably can build both shared and static. You didn't send the error, what is the actual problem, is it raylib, or maybe Glibc (if you compile on the latest Glibc you probably can not use it on older distros). What does `ldd` say for your binary?

Also, not sure why you are compiling main.c with GCC and then linking to stdc++. Is it a C or C++ project? The best way is to compile a fully static binary, Glibc will complain about that but the Musl toolchain actually recommends that you build the static binary. Such binary in theory should work on every distribution.

Anyway, doesn't look like a raylib issue, this is a general Linux question.

**Reply by 2gud4u** [reply]
Hello @gen2brain  yes i think its a raylib issue, here i cannot find libraylib.so.400 afther making a static compile on raylib, not saying its something wrong with raylib

[jorgecg@codigo-5 The XType]$ ldd xtype
	linux-vdso.so.1 (0x00007ffc457bf000)
	libraylib.so.400 => not found
	libGL.so.1 => /usr/lib/libGL.so.1 (0x00007fb547541000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007fb547459000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x00007fb547316000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007fb547000000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007fb546e19000)
	libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x00007fb54725c000)
	libGLX.so.0 => /usr/lib/libGLX.so.0 (0x00007fb546de7000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fb54761d000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007fb546dbc000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007fb54723c000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x00007fb546db7000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007fb546daf000)
[jorgecg@codigo-5 The XType]$

**Comment by gen2brain** [comment, marked as answer]
Well, that is not static, you linked against the shared raylib .so library, so any Linux OS would need to have raylib installed as a shared library in order to use your app/game. `.so` is a dynamic/shared library, `.a` or `archive` is a static library.

**Reply by 2gud4u** [reply]
Oh so i need to make a dynamic shared library ? I thought i done it in the past, but ok thank you for everything, thank you for clearing it all

**Reply by gen2brain** [reply]
No, you just need to install raylib as a static library, I believe the default `make` will do that for you. make install will put the raylib.h in include, and libraylib.a in /usr/local/lib (usually). If you only have a static library, the compiler will have no choice but to link to it, but if you have both, it will prefer the shared one.

**Reply by 2gud4u** [reply]
Now im confused, for what i know i have made a static library with git clone https://github.com/raysan5/raylib.git raylib
cd raylib/src/
make PLATFORM=PLATFORM_DESKTOP # To make the static version. So i dont know anymore whats the problem

**Reply by gen2brain** [reply]
Ok, that command built a static raylib library, but it is not installed on your system, it is still in that same local directory you cloned. Now, `make install` installs the library globally so it can be used, i.e you can link your app against it (compiler, GCC can use LDFLAGS or -L for locations where you have libraries, /usr/local/lib should be somewhere in defaults).
The problem is that you have `libraylib.so.400`, you don't want that, that is a shared library, remove it and just leave a libraylib.a. The goal is to not see raylib.so in ldd output. Not sure how you got that `.so`, but if you want to distribute binary, you don't want to link to shared library.

**Reply by 2gud4u** [reply]
Yes i have already run make install i had in the past raylib linux packages but i have removed them, could that be the cause ? if i do a locate libraylib.so this is my input
[jorgecg@codigo-5 src]$ locate libraylib.so
/home/jorgecg/Documentos/Proyectos Raylib/Raylib/raylib/build/raylib/libraylib.so
/home/jorgecg/Documentos/Proyectos Raylib/Raylib/raylib/build/raylib/libraylib.so.4.0.0
/home/jorgecg/Documentos/Proyectos Raylib/Raylib/raylib/build/raylib/libraylib.so.400
/home/jorgecg/Documentos/raylib/libraylib.so
/home/jorgecg/Documentos/raylib/libraylib.so.3.5.0
/home/jorgecg/Documentos/raylib/libraylib.so.350
/usr/lib/libraylib.so
/usr/lib/libraylib.so.4.2.0
/usr/lib/libraylib.so.420
/usr/local/lib/libraylib.so
/usr/local/lib/libraylib.so.3.5.0
/usr/local/lib/libraylib.so.3.7.0
/usr/local/lib/libraylib.so.350
/usr/local/lib/libraylib.so.370
/usr/local/lib/libraylib.so.4.0.0
/usr/local/lib/libraylib.so.400

**Reply by gen2brain** [reply]
Huh, yes, you have some mix there, and yes, that is the cause. You have raylib installed as a shared library, not just in /usr/local but also in /usr/lib, you should not do that. Your package manager should be the only one that installs something in /usr/lib. Your custom stuff you don't want there, and raylib if you did not install via package manager is custom stuff.

It is your system, you are free to remove them all. You probably used a CMake build system before this, and it installed a shared library on your system. My suggestion is to remove all of that, especially those in /usr/lib. Then just compile the static raylib library with make/make install and use that instead.

**Reply by 2gud4u** [reply]
Wow now it works the compile/link could i also now distribute my "exe" ???

**Reply by 2gud4u** [reply]
Hello, when i distribute my exe they tell me they got this error message.- 
./engine: /lib/x86_64-linux-gnu/[libm.so](http://libm.so/).6: version GLIBC_2.35' not found (required by ./engine) ./engine: /lib/x86_64-linux-gnu/libc.so.6: version GLIBC_2.33' not found (required by ./engine)
./engine: /lib/x86_64-linux-gnu/[libc.so](http://libc.so/).6: version `GLIBC_2.34' not found (required by ./engine)

**Reply by orcmid** [reply]
It seems you removed some library conflicts but you are still compiling an executable that requires a shared library.  

The only way to get a self-contained single-file executable is to rely only on static libraries and object files you have compiled.  

You probably don't even need a make file for this, but a Linux/GCC expert will need to assist you in that.  

**One suggestion:** Look at `build-linux.sh` in `raylib/projects/scripts`.  There are comments that will help.  Also `raylib/projects/scripts/README.md`. Start by copying `build-linux.sh` and `core_basic_window.c` into a folder of your own and use the script there.  If you can get a `core_basic_window` that works, see how to adapt that setup for a project of your own.
